### PR TITLE
CATROID-168 refactoring MediaDownloadService.

### PR DESCRIPTION
**Pull request by PlantinumX**
-MediaDownloadService is now a foreground service if > 26
	-ToastMessage is displayed in Activity
	-Deleted unused code in some parts.

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/transfers/project/ResultReceiverWrapper.kt [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import android.os.Handler
+import android.os.ResultReceiver
+
+class ResultReceiverWrapper(handler: Handler): ResultReceiver(handler) {
+    var resultReceiverWrapperInterface: ResultReceiverWrapperInterface? = null
+
+    override fun onReceiveResult(resultCode: Int, resultData: Bundle?) {
+        resultReceiverWrapperInterface?.onReceiveResult(resultCode, resultData)
+    }
+
```
Comment (thmq) [inline catroid/src/main/java/org/catrobat/catroid/transfers/project/ResultReceiverWrapper.kt]: unnecessary newline. now that some of us are really contributing kotlin code, we should really introduce some lint, checkstyle and pmd checks to find stuff like this.

File: catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java [hunk 2]
```diff
@@ -313,8 +318,26 @@ private boolean isWhatsappInstalled() {
...
+	@Override
+	public void onReceiveResult(int resultCode, Bundle resultData) {
+		long progress;
+		if (resultCode == Constants.UPDATE_DOWNLOAD_PROGRESS) {
+			progress = resultData.getLong(ProgressResponseBody.TAG_PROGRESS);
+			updateProgressDialog(progress);
+		} else if (resultCode == Constants.UPDATE_DOWNLOAD_SUCCESS) {
+			progress = resultData.getLong(ProgressResponseBody.TAG_PROGRESS);
+			updateProgressDialog(progress);
+			ToastUtil.showSuccess(WebViewActivity.this, "Download successful");
```
Comment (thmq) [inline catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java]: There is a hardcoded string `"Download successful"`. you really must not do that, please have a look at how to correctly provide strings that are visible on the UI.

File: catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java [hunk 3]
```diff
@@ -313,8 +318,26 @@ private boolean isWhatsappInstalled() {
...
+		} else {
+			dismissProgressDialog();
+			ToastUtil.showError(WebViewActivity.this, "Download failed");
```
Comment (thmq) [inline catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java]: same as for `"Download successful"`.

File: catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java [hunk 4]
```diff
@@ -313,8 +318,26 @@ private boolean isWhatsappInstalled() {
 		}
 	}
 
+	@Override
+	public void onReceiveResult(int resultCode, Bundle resultData) {
+		long progress;
+		if (resultCode == Constants.UPDATE_DOWNLOAD_PROGRESS) {
```
Comment (thmq) [inline catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java]: you could use a `switch (resultCode) { case Constants.UPDATE_DOWNLOAD_PROGRESS: ...}`. and please use static imports instead of quantified ones for imports from `Constants`. (more about static imports [here](https://docs.oracle.com/javase/1.5.0/docs/guide/language/static-import.html). -> but in short just remove the `Constants` quantifier and press alt + enter select static import and let android studio do its magic).

File: catroid/src/main/java/org/catrobat/catroid/utils/StatusBarNotificationManager.java [hunk 5]
```diff
@@ -57,7 +57,7 @@
 	private static final StatusBarNotificationManager INSTANCE = new StatusBarNotificationManager();
 	public static final String CHANNEL_ID = "pocket_code_notification_channel_id";
 
-	private int notificationId;
+	private int notificationId  = Constants.FOREGROUND_SERVICE;
```
Comment (thmq) [inline catroid/src/main/java/org/catrobat/catroid/utils/StatusBarNotificationManager.java]: please use a static import (see other comment about it).

File: catroid/src/main/java/org/catrobat/catroid/utils/StatusBarNotificationManager.java [hunk 6]
```diff
@@ -56,8 +57,7 @@
 
 	private static final StatusBarNotificationManager INSTANCE = new StatusBarNotificationManager();
 	public static final String CHANNEL_ID = "pocket_code_notification_channel_id";
-
-	private int notificationId;
+	private int notificationId = FOREGROUND_SERVICE;
```
Comment (LPeteR90) [inline catroid/src/main/java/org/catrobat/catroid/utils/StatusBarNotificationManager.java]: Why do we initialize notificationId with `FOREGROUND_SERVICE` here? What is the reason for doing that? In case this is done to avoid some unwanted behavior connected to the incrementing of notificationId when creating a new notifciation, we probably should consider a clearer approach or at least a clearer naming for this constant.

File: catroid/src/main/java/org/catrobat/catroid/transfers/MediaDownloadService.java [hunk 7]
```diff
@@ -24,85 +24,55 @@
...
-				public void run() {
-					ToastUtil.showSuccess(getBaseContext(), messageId);
-				}
-			});
-		}
+		String programName = intent.getStringExtra(PROGRAM_NAME);
+		final ResultReceiver receiver = intent.getParcelableExtra(RECEIVER);
+		StatusBarNotificationManager manager = StatusBarNotificationManager.getInstance();
+		final int notificationId = manager.createDownloadNotification(getApplicationContext(), programName);
+		startForeground(notificationId,
```
Comment (LPeteR90) [inline catroid/src/main/java/org/catrobat/catroid/transfers/MediaDownloadService.java]: Regarding user experience (which we briefly talked about some time in the past, I believe): We do already have a download dialog. Showing an additional notification does not really provide a better experience. 

So basically, the question is, if we need a foreground service at all here (I know, [https://jira.catrob.at/browse/CATROID-168](CATROID-168) is about turning this service into a foreground service, but since a dialog is shown to the user anyway, we might just keep this service as a background service. 

So basically this is a question for a PO (@wslany) about how we want the media download to behave. 

Related to this question: How does the ui behave in case a user navigates back from the WebViewActivity (is this even possible?) and the media download has not yet finished? This might only be testable on an emulator when setting the connection to a very low speed.

Comment (LPeteR90) [reply catroid/src/main/java/org/catrobat/catroid/transfers/MediaDownloadService.java]: Another note: I'd be ok with this not being a foreground service, since the implications would not be problematic here.

Comment (PlantinumX) [reply catroid/src/main/java/org/catrobat/catroid/transfers/MediaDownloadService.java]: First of all I agree with you about changing/(closing and creating ) a  ticket for backgorund service for the MediaDownloadService as we talked about it weeks ago. But why now??? We talked weeks before, we could have asked @wslany before opening the PR. So the work/time I invested would be less than now.... 
We talked in the retroperspective about the lack of  communication in the team -> so pls next time we should talk thorugh it before anyone in the team does an PR and afterwards the conclusion  is "hmm yeah   pls implement it different and not as it is written in the ticket".

Comment (LPeteR90) [reply catroid/src/main/java/org/catrobat/catroid/transfers/MediaDownloadService.java]: First off, sorry for the late response, I was on vacation until today. 

Of course you are right, tickets should not be changed drastically when already being in CR. This ticket was not very well defined, so this situation is on me. 

However, changing the implementation from a foreground service back to a background service is only removing one LOC (and about five LOC that are unused after that), so this should not be much work. So basically the decision if we want a foreground service (with the implied progress notification additional to the progress dialog) or a background service (ui stays the same, but service may be stopped in case of app is moved to background) can be made during PO review. 

Note that I really do like the rest of the implementation and hope it can be merged soon, but there has to be a decision (which not I, but a PO should make) on this issue first.
